File: SquirrelApplicationDelegate.m
#import "SquirrelApplicationDelegate.h"

#import <rime_api.h>
#import "SquirrelConfig.h"
#import "SquirrelPanel.h"
#import <UserNotifications/UserNotifications.h>

static NSString* const kRimeWikiURL = @"https://github.com/rime/home/wiki";

@implementation SquirrelApplicationDelegate

- (IBAction)deploy:(id)sender {
  NSLog(@"Start maintenance...");
  [self shutdownRime];
  [self startRimeWithFullCheck:YES];
  [self loadSettings];
}

- (IBAction)syncUserData:(id)sender {
  NSLog(@"Sync user data");
  rime_get_api()->sync_user_data();
}

- (IBAction)configure:(id)sender {
  [[NSWorkspace sharedWorkspace]
      openURL:[NSURL fileURLWithPath:@"~/Library/Rime/"
                                         .stringByExpandingTildeInPath
                         isDirectory:YES]];
}

- (IBAction)openWiki:(id)sender {
  [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:kRimeWikiURL]];
}

void show_message(const char* msg_text, const char* msg_id) {
  UNUserNotificationCenter* center =
      UNUserNotificationCenter.currentNotificationCenter;
  [center requestAuthorizationWithOptions:UNAuthorizationOptionAlert |
                                          UNAuthorizationOptionProvisional
                        completionHandler:^(BOOL granted, NSError* error) {
                          if (error) {
                            NSLog(@"User notification authorization error: %@",
                                  error.debugDescription);
                          }
                        }];
  [center getNotificationSettingsWithCompletionHandler:^(
              UNNotificationSettings* settings) {
    if ((settings.authorizationStatus == UNAuthorizationStatusAuthorized ||
         settings.authorizationStatus == UNAuthorizationStatusProvisional) &&
        (settings.alertSetting == UNNotificationSettingEnabled)) {
      UNMutableNotificationContent* content =
          [[UNMutableNotificationContent alloc] init];
      content.title = NSLocalizedString(@"Squirrel", nil);
      content.subtitle = NSLocalizedString(@(msg_text), nil);
      if (@available(macOS 12.0, *)) {
        content.interruptionLevel = UNNotificationInterruptionLevelActive;
      }
      UNNotificationRequest* request =
          [UNNotificationRequest requestWithIdentifier:@"SquirrelNotification"
                                               content:content
                                               trigger:nil];
      [center addNotificationRequest:request
               withCompletionHandler:^(NSError* error) {
                 if (error) {
                   NSLog(@"User notification request error: %@",
                         error.debugDescription);
                 }
               }];
    }
  }];
}

static void show_status_message(const char* msg_text_long,
                                const char* msg_text_short,
                                const char* msg_id) {
  SquirrelPanel* panel = NSApp.squirrelAppDelegate.panel;
  NSString* msgLong = msg_text_long ? @(msg_text_long) : nil;
  NSString* msgShort = msg_text_short ? @(msg_text_short) : nil;
  [panel updateStatusLong:msgLong statusShort:msgShort];
}

void notification_handler(void* context_object,
                          RimeSessionId session_id,
                          const char* message_type,
                          const char* message_value) {
  if (!strcmp(message_type, "deploy")) {
    if (!strcmp(message_value, "start")) {
      show_message("deploy_start", message_type);
    } else if (!strcmp(message_value, "success")) {
      show_message("deploy_success", message_type);
    } else if (!strcmp(message_value, "failure")) {
      show_message("deploy_failure", message_type);
    }
    return;
  }
  // off?
  SquirrelApplicationDelegate* app_delegate = (__bridge id)context_object;
  if (app_delegate && ![app_delegate enableNotifications]) {
    return;
  }
  // schema change
  if (!strcmp(message_type, "schema")) {
    const char* schema_name = strchr(message_value, '/');
    if (schema_name) {
      ++schema_name;
      show_status_message(schema_name, schema_name, message_type);
    }
    return;
  }
  // option change
  if (!strcmp(message_type, "option")) {
    Bool state = message_value[0] != '!';
    const char* option_name = message_value + !state;
    struct rime_string_slice_t state_label_long =
        rime_get_api()->get_state_label_abbreviated(session_id, option_name,
                                                    state, False);
    struct rime_string_slice_t state_label_short =
        rime_get_api()->get_state_label_abbreviated(session_id, option_name,
                                                    state, True);

    if (state_label_long.str || state_label_short.str) {
      const char* short_message =
          state_label_short.length < strlen(state_label_short.str)
              ? NULL
              : state_label_short.str;
      show_status_message(state_label_long.str, short_message, message_type);
    }
  }
}

- (void)setupRime {
  NSString* userDataDir = @"~/Library/Rime".stringByExpandingTildeInPath;
  NSFileManager* fileManager = [NSFileManager defaultManager];
  if (![fileManager fileExistsAtPath:userDataDir]) {
    if (![fileManager createDirectoryAtPath:userDataDir
                withIntermediateDirectories:YES
                                 attributes:nil
                                      error:nil]) {
      NSLog(@"Error creating user data directory: %@", userDataDir);
    }
  }
  rime_get_api()->set_notification_handler(notification_handler,
                                           (__bridge void*)(self));
  RIME_STRUCT(RimeTraits, squirrel_traits);
  squirrel_traits.shared_data_dir =
      [NSBundle mainBundle].sharedSupportPath.UTF8String;
  squirrel_traits.user_data_dir = userDataDir.UTF8String;
  squirrel_traits.distribution_code_name = "Squirrel";
  squirrel_traits.distribution_name = "鼠鬚管";
  squirrel_traits.distribution_version = [[[NSBundle mainBundle]
      objectForInfoDictionaryKey:(NSString*)kCFBundleVersionKey] UTF8String];
  squirrel_traits.app_name = "rime.squirrel";
  rime_get_api()->setup(&squirrel_traits);
}

- (void)startRimeWithFullCheck:(BOOL)fullCheck {
  NSLog(@"Initializing la rime...");
  rime_get_api()->initialize(NULL);
  // check for configuration updates
  if (rime_get_api()->start_maintenance((Bool)fullCheck)) {
    // update squirrel config
    rime_get_api()->deploy_config_file("squirrel.yaml", "config_version");
  }
}

- (void)shutdownRime {
  [_config close];
  rime_get_api()->finalize();
}

- (void)loadSettings {
  _config = [[SquirrelConfig alloc] init];
  if (![_config openBaseConfig]) {
    return;
  }

  _enableNotifications = ![[_config getString:@"show_notifications_when"]
      isEqualToString:@"never"];
  [self.panel loadConfig:_config forDarkMode:NO];
  if (@available(macOS 10.14, *)) {
    [self.panel loadConfig:_config forDarkMode:YES];
  }
}

- (void)loadSchemaSpecificSettings:(NSString*)schemaId {
  if (schemaId.length == 0 || [schemaId characterAtIndex:0] == '.') {
    return;
  }
  SquirrelConfig* schema = [[SquirrelConfig alloc] init];
  if ([schema openWithSchemaId:schemaId baseConfig:self.config] &&
      [schema hasSection:@"style"]) {
    [self.panel loadConfig:schema forDarkMode:NO];
  } else {
    [self.panel loadConfig:self.config forDarkMode:NO];
  }
  if (@available(macOS 10.14, *)) {
    if ([schema openWithSchemaId:schemaId baseConfig:self.config] &&
        [schema hasSection:@"style"]) {
      [self.panel loadConfig:schema forDarkMode:YES];
    } else {
      [self.panel loadConfig:self.config forDarkMode:YES];
    }
  }
  [schema close];
}

// prevent freezing the system
- (BOOL)problematicLaunchDetected {
  BOOL detected = NO;
  NSURL* logfile = [[NSURL fileURLWithPath:NSTemporaryDirectory()
                               isDirectory:YES]
      URLByAppendingPathComponent:@"squirrel_launch.dat"];
  // NSLog(@"[DEBUG] archive: %@", logfile);
  NSData* archive = [NSData dataWithContentsOfURL:logfile
                                          options:NSDataReadingUncached
                                            error:nil];
  if (archive) {
    NSDate* previousLaunch =
        [NSKeyedUnarchiver unarchivedObjectOfClass:NSDate.class
                                          fromData:archive
                                             error:nil];
    if (previousLaunch.timeIntervalSinceNow >= -2) {
      detected = YES;
    }
  }
  NSDate* now = [NSDate date];
  NSData* record = [NSKeyedArchiver archivedDataWithRootObject:now
                                         requiringSecureCoding:NO
                                                         error:nil];
  [record writeToURL:logfile atomically:NO];
  return detected;
}

- (void)workspaceWillPowerOff:(NSNotification*)aNotification {
  NSLog(@"Finalizing before logging out.");
  [self shutdownRime];
}

- (void)rimeNeedsReload:(NSNotification*)aNotification {
  NSLog(@"Reloading rime on demand.");
  [self deploy:nil];
}

- (void)rimeNeedsSync:(NSNotification*)aNotification {
  NSLog(@"Sync rime on demand.");
  [self syncUserData:nil];
}

- (NSApplicationTerminateReply)applicationShouldTerminate:
    (NSApplication*)sender {
  NSLog(@"Squirrel is quitting.");
  rime_get_api()->cleanup_all_sessions();
  return NSTerminateNow;
}

// add an awakeFromNib item so that we can set the action method.  Note that
// any menuItems without an action will be disabled when displayed in the Text
// Input Menu.
- (void)awakeFromNib {
  NSNotificationCenter* center =
      [NSWorkspace sharedWorkspace].notificationCenter;
  [center addObserver:self
             selector:@selector(workspaceWillPowerOff:)
                 name:NSWorkspaceWillPowerOffNotification
               object:nil];

  NSDistributedNotificationCenter* notifCenter =
      [NSDistributedNotificationCenter defaultCenter];
  [notifCenter addObserver:self
                  selector:@selector(rimeNeedsReload:)
                      name:@"SquirrelReloadNotification"
                    object:nil];

  [notifCenter addObserver:self
                  selector:@selector(rimeNeedsSync:)
                      name:@"SquirrelSyncNotification"
                    object:nil];
}

- (void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
  [[NSDistributedNotificationCenter defaultCenter] removeObserver:self];
  [_panel hide];
}

@end  // SquirrelApplicationDelegate

@implementation NSApplication (SquirrelApp)

- (SquirrelApplicationDelegate*)squirrelAppDelegate {
  return (SquirrelApplicationDelegate*)self.delegate;
}

@end

File: SquirrelApplicationDelegate.swift
//
//  SquirrelApplicationDelegate.swift
//  Squirrel
//
//  Created by Leo Liu on 5/6/24.
//

import UserNotifications
import Sparkle
import AppKit

class SquirrelApplicationDelegate: NSObject, NSApplicationDelegate {
  static let rimeWikiURL = URL(string: "https://github.com/rime/home/wiki")!
  
  let rimeAPI: RimeApi = rime_get_api().pointee
  var config: SquirrelConfig?
  var panel: SquirrelPanel?
  var enableNotifications = false
  let updateController = SPUStandardUpdaterController(startingUpdater: true, updaterDelegate: nil, userDriverDelegate: nil)
  
  func applicationWillFinishLaunching(_ notification: Notification) {
    addObservers()
  }
  
  func applicationWillTerminate(_ notification: Notification) {
    NotificationCenter.default.removeObserver(self)
    DistributedNotificationCenter.default().removeObserver(self)
    panel?.hide()
  }
  
  @objc func deploy() {
    print("Start maintenance...")
    self.shutdownRime()
    self.startRime(fullCheck: true)
    self.loadSettings()
  }
  
  @objc func syncUserData() {
    print("Sync user data")
    let _ = rimeAPI.sync_user_data()
  }
  
  @objc func configure() {
    let configURL = try! FileManager.default.url(for: .libraryDirectory, in: .userDomainMask, appropriateFor: nil, create: false).appendingPathComponent("Rime", isDirectory: true)
    NSWorkspace.shared.open(configURL)
  }
  
  @objc func checkForUpdates() {
    if updateController.updater.canCheckForUpdates {
      print("Checking for updates")
      updateController.updater.checkForUpdates()
    } else {
      print("Cannot check for updates")
    }
  }
  
  @objc func openWiki() {
    NSWorkspace.shared.open(Self.rimeWikiURL)
  }
  
  static func showMessage(msgText: String?, msgId: String?) {
    let center = UNUserNotificationCenter.current()
    center.requestAuthorization(options: [.alert, .provisional]) { granted, error in
      if let error = error {
        print("User notification authorization error: \(error.localizedDescription)")
      }
    }
    center.getNotificationSettings { settings in
      if (settings.authorizationStatus == .authorized || settings.authorizationStatus == .provisional) && settings.alertSetting == .enabled {
        let content = UNMutableNotificationContent()
        content.title = NSLocalizedString("Squirrel", comment: "")
        if let msgText = msgText {
          content.subtitle = msgText
        }
        content.interruptionLevel = .active
        let request = UNNotificationRequest(identifier: "SquirrelNotification", content: content, trigger: nil)
        center.add(request) { error in
          if let error = error {
            print("User notification request error: \(error.localizedDescription)")
          }
        }
      }
    }
  }
  
  func setupRime() {
    let userDataDir = try! FileManager.default.url(for: .libraryDirectory, in: .userDomainMask, appropriateFor: nil, create: false).appendingPathComponent("Rime", isDirectory: true)
    let fileManager = FileManager.default
    if !fileManager.fileExists(atPath: userDataDir.path()) {
      do {
        try fileManager.createDirectory(at: userDataDir, withIntermediateDirectories: true)
      } catch {
        print("Error creating user data directory: \(userDataDir.path())")
      }
    }
    let notification_handler: @convention(c) (UnsafeMutableRawPointer?, RimeSessionId, UnsafePointer<CChar>?, UnsafePointer<CChar>?) -> Void = notificationHandler
    let context_object = Unmanaged.passUnretained(self).toOpaque()
    rimeAPI.set_notification_handler(notification_handler, context_object)
    
    var squirrelTraits = RimeTraits.rimeStructInit()
    Bundle.main.sharedSupportPath!.withCString { cString in
      squirrelTraits.shared_data_dir = cString
    }
    userDataDir.path().withCString { cString in
      squirrelTraits.user_data_dir = cString
    }
    "Squirrel".withCString { cString in
      squirrelTraits.distribution_code_name = cString
    }
    "鼠鬚管".withCString { cString in
      squirrelTraits.distribution_name = cString
    }
    (Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String).withCString { cString in
      squirrelTraits.distribution_version = cString
    }
    "rime.squirrel".withCString { cString in
      squirrelTraits.app_name = cString
    }
    rimeAPI.setup(&squirrelTraits)
  }
  
  func startRime(fullCheck: Bool) {
    print("Initializing la rime...")
    rimeAPI.initialize(nil)
    // check for configuration updates
    if rimeAPI.start_maintenance(fullCheck) {
      // update squirrel config
      // print("[DEBUG] maintenance suceeds")
      let _ = rimeAPI.deploy_config_file("squirrel.yaml", "config_version")
    } else {
      // print("[DEBUG] maintenance fails")
    }
  }
  
  func loadSettings() {
    config = SquirrelConfig()
    if !config!.openBaseConfig() {
      return
    }
    
    enableNotifications = config!.getString("show_notifications_when") != "never"
    if let panel = panel, let config = self.config {
      panel.load(config: config, forDarkMode: false)
      panel.load(config: config, forDarkMode: true)
    }
  }
  
  func loadSettings(for schemaID: String) {
    if schemaID.count == 0 || schemaID.first == "." {
      return
    }
    let schema = SquirrelConfig()
    if let panel = panel, let config = self.config {
      if schema.open(schemaID: schemaID, baseConfig: config) && schema.has(section: "style") {
        panel.load(config: schema, forDarkMode: false)
        panel.load(config: schema, forDarkMode: true)
      } else {
        panel.load(config: config, forDarkMode: false)
        panel.load(config: config, forDarkMode: true)
      }
    }
    schema.close()
  }
  
  // prevent freezing the system
  func problematicLaunchDetected() -> Bool {
    var detected = false
    let logFile = FileManager.default.temporaryDirectory.appendingPathComponent("squirrel_launch.json", conformingTo: .json)
    //print("[DEBUG] archive: \(logFile)")
    do {
      let archive = try Data(contentsOf: logFile, options: [.uncached])
      let decoder = JSONDecoder()
      decoder.dateDecodingStrategy = .millisecondsSince1970
      let previousLaunch = try decoder.decode(Date.self, from: archive)
      if previousLaunch.timeIntervalSinceNow >= -2 {
        detected = true
      }
    } catch let error as NSError where error.domain == NSCocoaErrorDomain && error.code == NSFileReadNoSuchFileError {
      
    } catch {
      print("Error occurred during processing launch time archive: \(error.localizedDescription)")
      return detected
    }
    do {
      let encoder = JSONEncoder()
      encoder.dateEncodingStrategy = .millisecondsSince1970
      let record = try encoder.encode(Date.now)
      try record.write(to: logFile)
    } catch {
      print("Error occurred during saving launch time to archive: \(error.localizedDescription)")
    }
    return detected
  }
  
  // add an awakeFromNib item so that we can set the action method.  Note that
  // any menuItems without an action will be disabled when displayed in the Text
  // Input Menu.
  func addObservers() {
    let center = NSWorkspace.shared.notificationCenter
    center.addObserver(forName: NSWorkspace.willPowerOffNotification, object: nil, queue: nil, using: workspaceWillPowerOff)
    
    let notifCenter = DistributedNotificationCenter.default()
    notifCenter.addObserver(forName: .init("SquirrelReloadNotification"), object: nil, queue: nil, using: rimeNeedsReload)
    notifCenter.addObserver(forName: .init("SquirrelSyncNotification"), object: nil, queue: nil, using: rimeNeedsSync)
  }
  
  func applicationShouldTerminate(_ sender: NSApplication) -> NSApplication.TerminateReply {
    print("Squirrel is quitting.")
    rimeAPI.cleanup_all_sessions()
    return .terminateNow
  }

}

private func notificationHandler(contextObject: UnsafeMutableRawPointer?, sessionId: RimeSessionId, messageTypeC: UnsafePointer<CChar>?, messageValueC: UnsafePointer<CChar>?) {
  let delegate: SquirrelApplicationDelegate = Unmanaged<SquirrelApplicationDelegate>.fromOpaque(contextObject!).takeUnretainedValue()
  
  let messageType = messageTypeC.map { String(cString: $0) }
  let messageValue = messageValueC.map { String(cString: $0) }
  if messageType == "deploy" {
    switch messageValue {
    case "start":
      SquirrelApplicationDelegate.showMessage(msgText: NSLocalizedString("deploy_start", comment: ""), msgId: messageType)
    case "success":
      SquirrelApplicationDelegate.showMessage(msgText: NSLocalizedString("deploy_success", comment: ""), msgId: messageType)
    case "failure":
      SquirrelApplicationDelegate.showMessage(msgText: NSLocalizedString("deploy_failure", comment: ""), msgId: messageType)
    default:
      break
    }
    return
  }
  // off
  if !delegate.enableNotifications {
    return
  }
  // schema change
  if messageType == "schema", let messageValue = messageValue, let schemaName = try? /^[^\/]*\/(.*)$/.firstMatch(in: messageValue)?.output.1 {
    delegate.showStatusMessage(msgTextLong: String(schemaName), msgTextShort: String(schemaName))
    return
  }
  // option change
  if messageType == "option" {
    let state = messageValue?.first != "!"
    let optionName = if state {
      messageValue
    } else {
      String(messageValue![messageValue!.index(after: messageValue!.startIndex)...])
    }
    let stateLabelLong = delegate.rimeAPI.get_state_label_abbreviated(sessionId, optionName?.cString(using: .utf8), state, false)
    let stateLabelShort = delegate.rimeAPI.get_state_label_abbreviated(sessionId, optionName?.cString(using: .utf8), state, true)
    let longLabel = stateLabelLong.str.map { String(cString: $0) }
    let shortLabel = stateLabelShort.str.map { String(cString: $0) }
    delegate.showStatusMessage(msgTextLong: longLabel, msgTextShort: shortLabel)
  }
}

private extension SquirrelApplicationDelegate {
  func showStatusMessage(msgTextLong: String?, msgTextShort: String?) {
    panel?.updateStatus(long: msgTextLong ?? "", short: msgTextShort ?? "")
  }
  
  func shutdownRime() {
    config?.close()
    rimeAPI.finalize()
  }
  
  func workspaceWillPowerOff(notification: Notification) {
    print("Finalizing before logging out.")
    self.shutdownRime()
  }
  
  func rimeNeedsReload(notification: Notification) {
    print("Reloading rime on demand.")
    self.deploy()
  }
  
  func rimeNeedsSync(notification: Notification) {
    print("Sync rime on demand.")
    self.syncUserData()
  }
}

extension NSApplication {
  var squirrelAppDelegate: SquirrelApplicationDelegate {
    self.delegate as! SquirrelApplicationDelegate
  }
}
